File: solutions/prolog/difference-of-squares/difference_of_squares.pl
%! square_of_sum(+N, -Result)
%
% The square_of_sum/2 predicate succeeds with Result
% as square of sum from 1 to a given end number N.
square_of_sum(N, Result) :-
    Result is ((N+N*N)/2)**2.

%! sum_of_squares(+N, -Result)
%
% The sum_of_squares/2 predicate succeeds with Result
% as sum of squares from 1 to a given end number N.
sum_of_squares(N, Result) :-
    Result is (N+3*N**2+2*N**3)/6.

%! difference(+N, -Result)
%
% The difference/2 predicate succeeds with Result as the
% difference between sum of squares and square of sum.
difference(N, Result) :-
    square_of_sum(N, Result1),
    sum_of_squares(N, Result2),
    Result is Result1 - Result2.

File: solutions/clojure/difference-of-squares/src/difference_of_squares.clj
(ns difference-of-squares)

(defn sum-of-squares
  "Calculate sum of squares from 1 to a given end number
      y = (x * (x+1) * (2x+1)) / 6"
  [num]
  (/ (* num (inc num) (inc (* 2 num))) 6))

(defn square-of-sum
  "Calculate square of sum from 1 to a given end number
      y = (x * (x+1) / 2)^2"
  [num]
  (let [sum (/ (* num (inc num)) 2)] (* sum sum)))

(defn difference
  "Calculate difference between sum of squares and square of sum"
  [num]
  (- (square-of-sum num) (sum-of-squares num)))
